diff --git a/accessible/base/CacheConstants.h b/accessible/base/CacheConstants.h
--- a/accessible/base/CacheConstants.h
+++ b/accessible/base/CacheConstants.h
@@ -67,17 +67,17 @@ struct RelationData {
  * describe different kinds of relations, depending on the element they
  * originate on. For example, an <output> element's `for` attribute describes a
  * CONTROLLER_FOR relation, while the `for` attribute of a <label> describes a
  * LABEL_FOR relation. To ensure we process these attributes appropriately,
  * RelationData.mValidTag contains the atom for the tag this attribute/relation
  * type pairing is valid on. If the pairing is valid for all tag types, this
  * field is null.
  */
-static constexpr RelationData kRelationTypeAtoms[] = {
+inline constexpr RelationData kRelationTypeAtoms[] = {
     {nsGkAtoms::aria_labelledby, nullptr, RelationType::LABELLED_BY,
      RelationType::LABEL_FOR},
     {nsGkAtoms::_for, nsGkAtoms::label, RelationType::LABEL_FOR,
      RelationType::LABELLED_BY},
     {nsGkAtoms::aria_controls, nullptr, RelationType::CONTROLLER_FOR,
      RelationType::CONTROLLED_BY},
     {nsGkAtoms::_for, nsGkAtoms::output, RelationType::CONTROLLED_BY,
      RelationType::CONTROLLER_FOR},
