diff --git a/mfbt/moz.build b/mfbt/moz.build
--- a/mfbt/moz.build
+++ b/mfbt/moz.build
@@ -168,16 +168,17 @@ UNIFIED_SOURCES += [
     "FloatingPoint.cpp",
     "HashFunctions.cpp",
     "JSONWriter.cpp",
     "Poison.cpp",
     "RandomNum.cpp",
     "RefCounted.cpp",
     "SHA1.cpp",
     "TaggedAnonymousMemory.cpp",
+    "ToString.cpp",
     "UniquePtrExtensions.cpp",
     "Unused.cpp",
     "Utf8.cpp",
 ]
 
 if CONFIG["MOZ_BUILD_APP"] not in (
     "memory",
     "tools/update-programs",
