diff --git a/build/midl.py b/build/midl.py
--- a/build/midl.py
+++ b/build/midl.py
@@ -46,17 +46,21 @@ def search_path(paths, path):
 # Filter-out -std= flag from the preprocessor command, as we're not preprocessing
 # C or C++, and the command would fail with the flag.
 def filter_preprocessor(cmd):
     prev = None
     for arg in cmd:
         if arg == "-Xclang":
             prev = arg
             continue
-        if not arg.startswith("-std="):
+        if not (
+            arg.startswith("-std=")
+            or arg.startswith("-std:")
+            or arg.startswith("/std:")
+        ):
             if prev:
                 yield prev
             yield arg
         prev = None
 
 
 # Preprocess all the direct and indirect inputs of midl, and put all the
 # preprocessed inputs in the given `base` directory. Returns a tuple containing
