diff --git a/mozglue/misc/moz.build b/mozglue/misc/moz.build
--- a/mozglue/misc/moz.build
+++ b/mozglue/misc/moz.build
@@ -2,16 +2,17 @@
 # vim: set filetype=python:
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 FINAL_LIBRARY = "mozglue"
 
 EXPORTS.mozilla += [
+    "AutoMemory.h",
     "AutoProfilerLabel.h",
     "AwakeTimeStamp.h",
     "decimal/Decimal.h",
     "decimal/DoubleConversion.h",
     "IntegerPrintfMacros.h",
     "LoggingCore.h",
     "MmapFaultHandler.h",
     "Now.h",
@@ -23,31 +24,33 @@ EXPORTS.mozilla += [
     "Sprintf.h",
     "SSE.h",
     "StackWalk.h",
     "TimeStamp.h",
     "Uptime.h",
 ]
 
 EXPORTS.mozilla.glue += [
+    "AutoMemory.h",
     "Debug.h",
     "WinUtils.h",
 ]
 
 if CONFIG["OS_ARCH"] == "WINNT":
     EXPORTS.mozilla += [
         "GetKnownFolderPath.h",
         "PreXULSkeletonUI.h",
         "StackWalk_windows.h",
         "StackWalkThread.h",
         "TimeStamp_windows.h",
         "WindowsDpiAwareness.h",
     ]
 
 SOURCES += [
+    "AutoMemory.cpp",
     "AutoProfilerLabel.cpp",
     "AwakeTimeStamp.cpp",
     "Debug.cpp",
     "LoggingCore.cpp",
     "MmapFaultHandler.cpp",
     "Printf.cpp",
     "SIMD.cpp",
     "StackWalk.cpp",
