diff --git a/xpcom/base/nscore.h b/xpcom/base/nscore.h
--- a/xpcom/base/nscore.h
+++ b/xpcom/base/nscore.h
@@ -150,17 +150,17 @@
  * Note that we're also setting this for code coverage and pgo profile
  * generation, because both of those require atexit hooks, which won't fire
  * if we're using _exit. Bug 1555974 covers improving this.
  *
  */
 #ifndef NS_FREE_PERMANENT_DATA
 #  if defined(NS_BUILD_REFCNT_LOGGING) || defined(MOZ_VALGRIND) ||            \
       defined(MOZ_ASAN) || defined(MOZ_TSAN) || defined(MOZ_CODE_COVERAGE) || \
-      defined(MOZ_PROFILE_GENERATE) || defined(JS_STRUCTURED_SPEW)
+      defined(MOZ_PROFILE_GENERATE) || defined(MOZ_CS_PROFILE_GENERATE) || defined(JS_STRUCTURED_SPEW)
 #    define NS_FREE_PERMANENT_DATA
 #  endif
 #endif
 
 /**
  * NS_NO_VTABLE is emitted by xpidl in interface declarations whenever
  * xpidl can determine that the interface can't contain a constructor.
  * This results in some space savings and possible runtime savings -
