diff --git a/security/sandbox/common/SandboxSettings.cpp b/security/sandbox/common/SandboxSettings.cpp
--- a/security/sandbox/common/SandboxSettings.cpp
+++ b/security/sandbox/common/SandboxSettings.cpp
@@ -216,17 +216,17 @@ int GetEffectiveSocketProcessSandboxLeve
 
   return level;
 }
 
 int GetEffectiveGpuSandboxLevel() {
   return StaticPrefs::security_sandbox_gpu_level();
 }
 
-#if defined(MOZ_PROFILE_GENERATE)
+#if defined(MOZ_PROFILE_GENERATE) || defined(MOZ_CS_PROFILE_GENERATE)
 // It should only be allowed on instrumented builds, never on production
 // builds.
 #  if defined(XP_WIN)
 bool GetLlvmProfileDir(std::wstring& parentPath) {
   bool rv = false;
   if (const wchar_t* llvmProfileFileEnv = _wgetenv(L"LLVM_PROFILE_FILE")) {
     std::wstring llvmProfileDir(llvmProfileFileEnv);
     const size_t found = llvmProfileDir.find_last_of(L"/\\");
@@ -247,17 +247,17 @@ bool GetLlvmProfileDir(std::string& pare
     if (found != std::string::npos) {
       parentPath = llvmProfileDir.substr(0, found);
       rv = true;
     }
   }
   return rv;
 }
 #  endif  // defined(XP_WIN)
-#endif    // defined(MOZ_PROFILE_GENERATE
+#endif    // defined(MOZ_PROFILE_GENERATE) || defined(MOZ_CS_PROFILE_GENERATE)
 
 #if defined(XP_MACOSX)
 int ClampFlashSandboxLevel(const int aLevel) {
   const int minLevel = 0;
   const int maxLevel = 3;
 
   if (aLevel < minLevel) {
     return minLevel;
