diff --git a/gfx/skia/skia/src/core/SkOpts.h b/gfx/skia/skia/src/core/SkOpts.h
--- a/gfx/skia/skia/src/core/SkOpts.h
+++ b/gfx/skia/skia/src/core/SkOpts.h
@@ -68,18 +68,20 @@ namespace SkOpts {
     // Call to replace pointers to portable functions with pointers to CPU-specific functions.
     // Thread-safe and idempotent.
     // Called by SkGraphics::Init().
     void Init();
 
     // We can't necessarily express the type of SkRasterPipeline stage functions here,
     // so we just use this void(*)(void) as a stand-in.
     using StageFn = void(*)(void);
-    extern StageFn ops_highp[kNumRasterPipelineHighpOps], just_return_highp;
-    extern StageFn ops_lowp [kNumRasterPipelineLowpOps ], just_return_lowp;
+    extern const StageFn* ops_highp;
+    extern const StageFn* ops_lowp;
+    extern StageFn just_return_highp;
+    extern StageFn just_return_lowp;
 
     extern void (*start_pipeline_highp)(size_t,size_t,size_t,size_t, SkRasterPipelineStage*,
                                         SkSpan<SkRasterPipelineContexts::MemoryCtxPatch>,
                                         uint8_t*);
     extern void (*start_pipeline_lowp )(size_t,size_t,size_t,size_t, SkRasterPipelineStage*,
                                         SkSpan<SkRasterPipelineContexts::MemoryCtxPatch>,
                                         uint8_t*);
 
