diff --git a/dom/media/webrtc/transport/third_party/nICEr/src/net/transport_addr.h b/dom/media/webrtc/transport/third_party/nICEr/src/net/transport_addr.h
--- a/dom/media/webrtc/transport/third_party/nICEr/src/net/transport_addr.h
+++ b/dom/media/webrtc/transport/third_party/nICEr/src/net/transport_addr.h
@@ -34,31 +34,35 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 
 #ifndef _transport_addr_h
 #define _transport_addr_h
 
 #include <stdbool.h>
 #include <sys/types.h>
 #ifdef WIN32
 
+#ifndef MOZ_UNIFIED_BUILD
 // FIXME: This is dangerous, but exactly the pattern used in
 // nrappkit/src/port/win32/include/csi_platform.h
 // Not good because INT8 are typedefed to different values in
 // <winsock2.h> and <r_types.h>.
 // {
 typedef unsigned char UBLAH_IGNORE_ME_PLEASE;
 typedef signed char BLAH_IGNORE_ME_PLEASE;
 #define UINT8 UBLAH_IGNORE_ME_PLEASE
 #define INT8 BLAH_IGNORE_ME_PLEASE
+#endif
 #include <winsock2.h>
+#ifndef MOZ_UNIFIED_BUILD
 #undef UINT8
 #undef INT8
 #include <r_types.h>
 
 // }
+#endif
 
 #include <ws2tcpip.h>
 #else
 #include <sys/socket.h>
 #include <netinet/in.h>
 #endif
 
 #include "r_types.h"
